File: sdk/js/src/modules/storage.ts
// sdk/js/src/modules/storage.ts

import { BaseFetch } from "../utils/fetch";
import type {
  StorageUploadOptions,
  StorageListOptions,
  StorageDownloadOptions,
} from "../types/filters";
import type {
  UploadResult,
  FileMeta,
  DeleteFileResult,
  DownloadUrlResult,
} from "../types/index";

// ─── Raw API shapes ───────────────────────────────────────────────────────────

interface UploadUrlResponse {
  upload_url: string;
  file_url: string;
  filename: string;
  bucket: string;
}

interface DownloadUrlResponse {
  url: string;
  path: string;
}

interface DeleteFileResponse {
  deleted: boolean;
  key: string;
}

// ─── BucketBuilder ────────────────────────────────────────────────────────────

/**
 * Scoped to a single bucket. Returned by `baas.storage("bucket-name")`.
 */
export class BucketBuilder {
  private http: BaseFetch;
  private projectId: string;
  private bucketName: string;

  constructor(http: BaseFetch, projectId: string, bucket: string) {
    this.http = http;
    this.projectId = projectId;
    this.bucketName = bucket;
  }

  /**
   * Get a presigned PUT URL for a direct client-to-storage upload.
   * The file is never proxied through the API server.
   *
   * @example
   * const { uploadUrl, fileUrl } = await baas.storage("avatars").upload({
   *   filename: "profile.jpg",
   *   contentType: "image/jpeg",
   * })
   * await fetch(uploadUrl, { method: "PUT", body: fileBlob })
   * // fileUrl is now the permanent/presigned download URL
   */
  async upload(options: StorageUploadOptions): Promise<UploadResult> {
    const res = await this.http.post<UploadUrlResponse>(
      `/v1/storage/${this.projectId}/${this.bucketName}/upload`,
      {
        filename: options.filename,
        content_type: options.contentType,
        expires_in: options.expiresIn ?? 3600,
      }
    );
    return {
      uploadUrl: res.data.upload_url,
      fileUrl: res.data.file_url,
      filename: res.data.filename,
      bucket: res.data.bucket,
    };
  }

  /**
   * List files in this bucket, optionally filtered by a key prefix.
   *
   * @example
   * const files = await baas.storage("avatars").listFiles()
   * const docs = await baas.storage("docs").listFiles({ prefix: "reports/", limit: 50 })
   */
  async listFiles(options?: StorageListOptions): Promise<FileMeta[]> {
    const query: Record<string, string | number | undefined> = {};
    if (options?.prefix) query["prefix"] = options.prefix;
    if (options?.limit) query["limit"] = options.limit;

    const res = await this.http.get<FileMeta[]>(
      `/v1/storage/${this.projectId}/${this.bucketName}/files`,
      query
    );
    return res.data;
  }

  /**
   * Delete a file by its key/path.
   *
   * @example
   * await baas.storage("avatars").deleteFile("profile.jpg")
   * await baas.storage("docs").deleteFile("reports/q3.pdf")
   */
  async deleteFile(path: string): Promise<DeleteFileResult> {
    const res = await this.http.delete<DeleteFileResponse>(
      `/v1/storage/${this.projectId}/${this.bucketName}/${path}`
    );
    return { deleted: res.data.deleted, key: res.data.key };
  }

  /**
   * Get a presigned download URL for a file.
   * The URL is valid for `expiresIn` seconds (default 3600).
   *
   * @example
   * const { url } = await baas.storage("avatars").getDownloadUrl("profile.jpg")
   * const { url } = await baas.storage("docs").getDownloadUrl("report.pdf", { expiresIn: 300 })
   */
  async getDownloadUrl(
    path: string,
    options?: StorageDownloadOptions
  ): Promise<DownloadUrlResult> {
    const query: Record<string, number | undefined> = {};
    if (options?.expiresIn) query["expires_in"] = options.expiresIn;

    const res = await this.http.get<DownloadUrlResponse>(
      `/v1/storage/${this.projectId}/${this.bucketName}/${path}/url`,
      query
    );
    return { url: res.data.url, path: res.data.path };
  }
}

// ─── StorageModule ────────────────────────────────────────────────────────────

export class StorageModule {
  private http: BaseFetch;
  private projectId: string;

  constructor(http: BaseFetch, projectId: string) {
    this.http = http;
    this.projectId = projectId;
  }

  /**
   * Returns a BucketBuilder scoped to the given bucket name.
   *
   * @example
   * baas.storage("avatars").upload({ filename: "pic.jpg", contentType: "image/jpeg" })
   * baas.storage("docs").listFiles({ prefix: "reports/" })
   * baas.storage("avatars").deleteFile("old-pic.jpg")
   * baas.storage("docs").getDownloadUrl("report.pdf")
   */
  bucket(name: string): BucketBuilder {
    return new BucketBuilder(this.http, this.projectId, name);
  }
}

File: sdk/python/src/baas/modules/storage.py
# sdk/python/src/yourbaas/modules/storage.py
from __future__ import annotations

from typing import Any

from baas.types import FileMeta, PresignedUpload
from baas.utils.http import HttpClient


class BucketBuilder:
    """
    MinIO storage operations for a specific bucket.

    Files are NEVER proxied through FastAPI — all uploads/downloads use
    presigned URLs that go directly to MinIO.

    Usage:
        # Get a presigned upload URL
        result = await baas.storage("avatars").upload(
            filename="profile.jpg",
            content_type="image/jpeg",
        )
        # Then PUT directly to result.upload_url using any HTTP client

        # List files
        files = await baas.storage("avatars").list_files()
        files = await baas.storage("avatars").list_files(prefix="2024/")

        # Get a download URL
        url = await baas.storage("avatars").get_download_url("path/to/file.jpg")

        # Delete a file
        await baas.storage("avatars").delete_file("path/to/file.jpg")
    """

    def __init__(self, http: HttpClient, project_id: str, bucket: str) -> None:
        self._http = http
        self._project_id = project_id
        self._bucket = bucket

    def _base(self) -> str:
        return f"/v1/storage/{self._project_id}/{self._bucket}"

    async def upload(
        self,
        *,
        filename: str,
        content_type: str,
        expires_in: int = 3600,
    ) -> PresignedUpload:
        """
        Request a presigned PUT URL for direct client upload.

        Returns a PresignedUpload with:
        - upload_url: PUT this URL directly from the client
        - file_url: The final publicly accessible download URL
        - key: The storage key (path) of the file
        """
        result = await self._http.post(
            f"{self._base()}/upload",
            json_body={
                "filename": filename,
                "content_type": content_type,
                "expires_in": expires_in,
            },
        )
        if isinstance(result, dict):
            return PresignedUpload(
                upload_url=result.get("upload_url", ""),
                file_url=result.get("file_url", ""),
                key=result.get("filename", filename),
                expires_in=expires_in,
            )
        raise ValueError(f"Unexpected response from storage upload: {result}")

    async def list_files(
        self,
        *,
        prefix: str | None = None,
        limit: int = 100,
    ) -> list[FileMeta]:
        """List files in this bucket, optionally filtered by prefix."""
        params: dict[str, Any] = {"limit": limit}
        if prefix:
            params["prefix"] = prefix

        result = await self._http.get(f"{self._base()}/files", params=params)
        items = result if isinstance(result, list) else []
        return [
            FileMeta(
                key=item.get("key", ""),
                size=item.get("size", 0),
                last_modified=item.get("last_modified"),
                content_type=item.get("content_type"),
                etag=item.get("etag"),
            )
            for item in items
        ]

    async def delete_file(self, file_path: str) -> bool:
        """Delete a file by its storage path/key. Returns True on success."""
        from baas.utils.errors import BaasError
        try:
            result = await self._http.delete(f"{self._base()}/{file_path}")
            if isinstance(result, dict):
                return result.get("deleted", False)
            return True
        except BaasError as e:
            if e.status == 404:
                return False
            raise

    async def get_download_url(
        self,
        file_path: str,
        *,
        expires_in: int = 3600,
    ) -> str:
        """
        Get a presigned download URL for a file.
        The URL is valid for expires_in seconds (default 1 hour).
        """
        result = await self._http.get(
            f"{self._base()}/{file_path}/url",
            params={"expires_in": expires_in},
        )
        if isinstance(result, dict):
            return result.get("url", "")
        return str(result)


class StorageModule:
    """
    Entry point for storage (MinIO) operations.

    baas.storage("bucket_name") → BucketBuilder
    """

    def __init__(self, http: HttpClient, project_id: str) -> None:
        self._http = http
        self._project_id = project_id

    def __call__(self, bucket: str) -> BucketBuilder:
        """Return a BucketBuilder for the given bucket."""
        return BucketBuilder(self._http, self._project_id, bucket)
